test/integration/master/kube_apiserver_test.go

```diff
@@ -618,3 +638,130 @@ func TestMultiMasterNodePortAllocation(t *testing.T) {
+	var clientAPIServers []*kubernetes.Clientset
+	etcd := framework.SharedEtcd()
+
+	instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+		DisableStorageCleanup: true,
+	}
+
+	// cleanup the registry storage
+	defer registry.CleanupStorage()
+
+	// create 1 api servers and 1 clients: cannot test with multiple ones because then there would be a conflict as each one has a different bind port configured
+	for i := 0; i < 1; i++ {
+		// start master count api server
+		t.Logf("starting api server: %d", i)
+		server := kubeapiservertesting.StartTestServerOrDie(t, instanceOptions, []string{
+			"--endpoint-reconciler-type", "lease",
+			"--advertise-address", fmt.Sprintf("10.0.1.%v", i+1),
+		}, etcd)
+		kubeAPIServers = append(kubeAPIServers, server)
+
+		// verify kube API servers have registered and create a client
+		if err := wait.PollImmediate(3*time.Second, 2*time.Minute, func() (bool, error) {
+			client, err := kubernetes.NewForConfig(kubeAPIServers[i].ClientConfig)
+			if err != nil {
+				t.Logf("create client error: %v", err)
+				return false, nil
+			}
+			clientAPIServers = append(clientAPIServers, client)
+			services, err := client.CoreV1().Services("default").Get(context.TODO(), "kubernetes", metav1.GetOptions{})
+			if err != nil {
+				t.Logf("error fetching services: %v", err)
+				return false, nil
+			}
+			if !reflect.DeepEqual([]int{server.ServerOpts.SecureServing.BindPort}, getServiceTargetPorts(services)) {
+				t.Logf("error comparing target port in services with bind port set: Bind port %v vs targetPort %v", server.ServerOpts.SecureServing.BindPort, getServiceTargetPorts(services))
+				return false, nil
+			}
+			endpoints, err := client.CoreV1().Endpoints("default").Get(context.TODO(), "kubernetes", metav1.GetOptions{})
+			if err != nil {
+				t.Logf("error fetching endpoints: %v", err)
+				return false, nil
+			}
+			if !reflect.DeepEqual([]int{server.ServerOpts.SecureServing.BindPort}, getEndpointPorts(endpoints)) {
+				t.Logf("error comparing endpoints port with advertise-port: %v", getEndpointPorts(endpoints))
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			t.Fatalf("did not find only lease endpoints: %v", err)
+		}
+	}
+
+	// shutdown the api servers
+	for _, server := range kubeAPIServers {
+		server.TearDownFn()
+	}
+
+}
+
+// TestMultiMasterAdvertisePort tests that when using the --advertise-port the targetPort in the kubernetes.default.svc server
+// and also as port in the kubernetes.default.svc endpoints and it's the value specified there
+func TestMultiMasterAdvertisePort(t *testing.T) {
+	var kubeAPIServers []*kubeapiservertesting.TestServer
+	var clientAPIServers []*kubernetes.Clientset
+	etcd := framework.SharedEtcd()
+
+	instanceOptions := &kubeapiservertesting.TestServerInstanceOptions{
+		DisableStorageCleanup: true,
+	}
+
+	// cleanup the registry storage
+	defer registry.CleanupStorage()
+
+	// create 2 api servers and 2 clients
+	for i := 0; i < 2; i++ {
```

Comment:
I don't get why we need to start two apiservers in the test. Can you explain?

Reply:
I agree, I don't think we do for this particular test (though @aojea's comments below wrt multiple apiservers _does_ need to be addressed, and that will require multiple).